atlasdb-config/src/main/java/com/palantir/atlasdb/factory/timelock/TimestampCorroboratingTimelockService.java

```diff
@@ -0,0 +1,94 @@
+
+import com.palantir.lock.v2.AutoDelegate_TimelockService;
+import com.palantir.lock.v2.IdentifiedTimeLockRequest;
+import com.palantir.lock.v2.StartAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.timestamp.TimestampRange;
+
+/**
+ * A timelock service decorator for introducing runtime validity checks on received timestamps.
+ */
+public class TimestampCorroboratingTimelockService implements AutoDelegate_TimelockService {
+    static final String CLOCKS_WENT_BACKWARDS_MESSAGE =
+            "Expected timestamp to be greater than %s, but a fresh timestamp was %s!";
+
+    private final TimelockService delegate;
+    private volatile long lowerBound = Long.MIN_VALUE;
+
+    private TimestampCorroboratingTimelockService(TimelockService delegate) {
+        this.delegate = delegate;
+    }
+
+    public static TimelockService create(TimelockService delegate) {
+        return new TimestampCorroboratingTimelockService(delegate);
+    }
+
+    @Override
+    public TimelockService delegate() {
+        return delegate;
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return checkAndUpdate(delegate::getFreshTimestamp, x -> x, x -> x);
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return checkAndUpdate(() -> delegate.getFreshTimestamps(numTimestampsRequested),
+                TimestampRange::getLowerBound,
+                TimestampRange::getUpperBound);
+    }
+
+    @Override
+    public StartAtlasDbTransactionResponse startAtlasDbTransaction(IdentifiedTimeLockRequest request) {
+        return checkAndUpdate(() -> delegate.startAtlasDbTransaction(request),
+                StartAtlasDbTransactionResponse::freshTimestamp,
+                StartAtlasDbTransactionResponse::freshTimestamp);
+    }
+
+    private <T> T checkAndUpdate(Supplier<T> timestampContainerSupplier,
+            ToLongFunction<T> lowerBoundExtractor,
+            ToLongFunction<T> upperBoundExtractor) {
+        long threadLocalLowerBound = lowerBound;
+        T timestampContainer = timestampContainerSupplier.get();
+
+        checkTimestamp(threadLocalLowerBound, lowerBoundExtractor.applyAsLong(timestampContainer));
+        updateLowerBound(upperBoundExtractor.applyAsLong(timestampContainer));
+        return timestampContainer;
+    }
+
+    private void checkTimestamp(long timestampLowerBound, long freshTimestamp) {
+        if (freshTimestamp <= lowerBound) {
+            throw clocksWentBackwards(timestampLowerBound, freshTimestamp);
+        }
+    }
+
+    private synchronized void updateLowerBound(long freshTimestamp) {
+        lowerBound = Math.max(freshTimestamp, lowerBound);
+    }
+
+    private AssertionError clocksWentBackwards(long timestampLowerBound, long freshTimestamp) {
+        return new AssertionError(String.format(CLOCKS_WENT_BACKWARDS_MESSAGE, timestampLowerBound, freshTimestamp));
```

Comment:
this isn't really an AssertionError; this has to be an IllegalStateException, I think. Like, Assertion error implies bug in code; this could be bug in deployment.

Reply:
in fact; make it a SafeIllegalStateException and add safe args with the params you care about

Reply:
(also, thinking about it - do we still need the install time check?)

Reply:
we should merge the install check into this one; instead of initializing at Long.MIN_VALUE, initialize with the punch table one

Reply:
I will address this with a separate pr shortly, as I am not sure if the end result will be cleaner than the current state.

Reply:
#3594